SNV
===

Identifies the nucleotides that overlap each position in the BAM file. All reads formed by a consensus less than the
min_read_per_uid threshold are ignored. The type of each nucleotide at a position are assigned as follows:

    1. If a read is in duplex, its' bases are assigned as duplex supported bases (D). Otherwise, they are assigned as singletons (S)
    2. For each stand (+ and -), if a consensus read was generated by at least strong_supported_base_threshold number of original reads, that base is assigned as a strong base (P and N). Otherwise, they are assigned as weak bases (p and n).


Run Using
^^^^^^^^^

::

    produse snv

or

::

    python /path/to/ProDuSe/ProDuSe/snv.py

Parameters
^^^^^^^^^^

    :-c --config:
        An optional configuration file which can specify any of the arguments described below
    :-i --input:
        An input BAM containing collapsed reads. This file should be sorted by read position.
    :-o --output:
        Path and name of an output VCF file containing candidate variants.
    :-s --molecule_stats:
        Path and file name of an output tsv file containing position depth and molecule abundance information.
    :-r --reference:
        Reference genome, in FASTA format. An index should be present in the same directory
    :-tb --target_bed:
        A tab-delineated BED3 file in which to restrict variant calling.
    :-vaft --variant_allele_fraction_threshold:
        Minimum variant allele fraction to call a variant on each strand.
    :-mo --min_molecules:
        Total molecules (i.e. actual depth) required to call a variant at this position. This threshold should be adjusted if you are restriction to sites expected to be mutated.
    :-mum --mutant_molecules:
        Minimum number of supporting molecules on both strands required to call a variant.
    :-mrpu --min_reads_per_uid:
        Each consensus read must have been formed from at least this number of original reads to use this read in variant calling.
    :-ssbt --strong_supported_base_threshold:
        Bases from a consensus read collapsed from at this number of original reads will be classified as a strong supporting base.
    :-eds --enforce_dual_strand:
        Only variants with support on both the positive and negative strand will be called.
    :-mq --min_qual:
        Bases from weak supporting molecules with less than this quality threshold will be treated as an "N".


